File: Loenn/entities/void_tendril.lua
local void_tendril = {}

void_tendril.name = "DesoloZantas/VoidTendril"
void_tendril.depth = 100
void_tendril.nodeLimits = {0, -1}
void_tendril.placements = {
    {
        name = "Void Tendril",
        data = {
            damage = 1,
            swaySpeed = 1.0,
            phaseableOnDash = true,
            phaseWindow = 0.3
        }
    },
    {
        name = "Void Tendril (Impassable)",
        data = {
            damage = 1,
            swaySpeed = 1.0,
            phaseableOnDash = false,
            phaseWindow = 0.0
        }
    }
}

void_tendril.fieldInformation = {
    damage = {
        fieldType = "integer",
        minimumValue = 1,
        maximumValue = 5
    },
    swaySpeed = {
        fieldType = "number",
        minimumValue = 0.1,
        maximumValue = 5.0
    },
    phaseWindow = {
        fieldType = "number",
        minimumValue = 0.1,
        maximumValue = 1.0
    }
}

function void_tendril.sprite(room, entity)
    local sprites = {}
    local x, y = entity.x, entity.y
    local height = 48
    
    local drawableRectangle = require("structs.drawable_rectangle")
    
    -- Draw tendril body
    for i = 0, height - 4, 4 do
        local width = 8 - (i / height) * 4  -- Tapers toward top
        local alpha = 0.6 + (i / height) * 0.3
        local rect = drawableRectangle.fromRectangle("fill", x - width/2, y + i, width, 4, {0.4, 0.0, 0.4, alpha})
        for _, sprite in ipairs(rect:getDrawableSprite()) do
            table.insert(sprites, sprite)
        end
    end
    
    return sprites
end

function void_tendril.nodeSprite(room, entity, node, nodeIndex)
    local sprites = {}
    local drawableRectangle = require("structs.drawable_rectangle")
    local rect = drawableRectangle.fromRectangle("fill", node.x - 4, node.y, 8, 8, {0.6, 0.0, 0.6, 0.5})
    
    for _, sprite in ipairs(rect:getDrawableSprite()) do
        table.insert(sprites, sprite)
    end
    
    return sprites
end

return void_tendril

File: Source/Core/Entities/VoidTendril.cs
using System;
using System.Collections.Generic;
using Celeste;
using Microsoft.Xna.Framework;
using Monocle;

namespace Celeste.Mod.DesoloZatnas.Core.Entities
{
    /// <summary>
    /// Dangerous tentacle from the void
    /// Can be phased through with a well-timed dash
    /// </summary>
    public class VoidTendril : Entity
    {
        private int damage;
        private float swaySpeed;
        private bool phaseableOnDash;
        private float phaseWindow;
        
        private float swayTimer = 0f;
        private float swayOffset = 0f;
        private List<Vector2> nodes;
        
        private bool playerPhasing = false;
        private float phaseTimer = 0f;
        
        private Color tendrilColor;
        private Color coreColor;
        
        public VoidTendril(EntityData data, Vector2 offset)
            : base(data.Position + offset)
        {
            damage = data.Int("damage", 1);
            swaySpeed = data.Float("swaySpeed", 1.0f);
            phaseableOnDash = data.Bool("phaseableOnDash", true);
            phaseWindow = data.Float("phaseWindow", 0.3f);
            
            nodes = new List<Vector2>();
            nodes.Add(Position);
            foreach (var node in data.Nodes)
            {
                nodes.Add(node + offset);
            }
            
            Collider = new Hitbox(8f, 48f, -4f, 0f);
            Add(new PlayerCollider(OnPlayer));
            
            tendrilColor = new Color(80, 0, 80);
            coreColor = new Color(160, 0, 160);
            
            Depth = 100;
        }
        
        public override void Update()
        {
            base.Update();
            
            swayTimer += Engine.DeltaTime * swaySpeed;
            swayOffset = (float)Math.Sin(swayTimer) * 8f;
            
            // Update collision position with sway
            Collider.Position = new Vector2(-4f + swayOffset, 0f);
            
            if (playerPhasing)
            {
                phaseTimer -= Engine.DeltaTime;
                if (phaseTimer <= 0f)
                {
                    playerPhasing = false;
                }
            }
            
            // Check for player dash to start phasing
            if (phaseableOnDash && !playerPhasing)
            {
                Player player = Scene?.Tracker.GetEntity<Player>();
                if (player != null && player.StateMachine.State == Player.StDash)
                {
                    float distSq = Vector2.DistanceSquared(player.Center, Center);
                    if (distSq < 50 * 50)
                    {
                        StartPhasing();
                    }
                }
            }
        }
        
        private void StartPhasing()
        {
            playerPhasing = true;
            phaseTimer = phaseWindow;
            
            Audio.Play("event:/game/general/fallblock_shake", Position);
        }
        
        private void OnPlayer(Player player)
        {
            if (playerPhasing)
            {
                // Player successfully phased through
                return;
            }
            
            // Deal damage
            player.Die(Vector2.UnitX * (player.X < X ? -1f : 1f));
        }
        
        public override void Render()
        {
            float alpha = playerPhasing ? 0.3f : 1f;
            
            // Draw tendril body
            float height = 48f;
            for (float y = 0; y < height; y += 4f)
            {
                float width = 8f - (y / height) * 4f; // Tapers toward top
                float xOffset = swayOffset * (y / height); // More sway at top
                
                Color color = Color.Lerp(tendrilColor, coreColor, y / height) * alpha;
                Draw.Rect(X - width / 2f + xOffset, Y + y, width, 4f, color);
            }
            
            // Draw pulsing core
            float pulse = 0.7f + (float)Math.Sin(swayTimer * 3f) * 0.3f;
            Draw.Circle(Position + new Vector2(swayOffset * 0.5f, height * 0.3f), 4f * pulse, coreColor * alpha, 8);
            
            // Draw phase indicator if phaseable
            if (phaseableOnDash && !playerPhasing)
            {
                float indicatorAlpha = 0.3f + (float)Math.Sin(swayTimer * 2f) * 0.2f;
                Draw.Circle(Position + new Vector2(0, height * 0.5f), 16f, Color.Cyan * indicatorAlpha, 16);
            }
        }
    }
}
